File: backend/app/routes/public.py
from fastapi import APIRouter, Depends
from fastapi.responses import JSONResponse
from sqlalchemy.orm import Session
from sqlalchemy import func
from typing import List, Optional
from datetime import datetime

from app.core.database import get_db
from app.models.mission import Mission
from app.models.reading import Reading
from app.models.hotspot import Hotspot
from app.schemas.public import (
    PublicOverview,
    PublicMission,
    PublicMissionList,
    PublicMissionDetail,
    PublicMapPoint,
    PublicHotspot,
    PublicHotspotList,
    PublicMissionMapData,
    PublicTrendPoint,
    PublicTrendData,
    PublicAIQueryRequest,
    PublicAIResponse,
)

# Reuse RAG pipeline components
from app.services.ai_router import detect_intent_and_extract_params
from app.services.rag.context_builder import get_context_builder
from app.services.rag.grounding import get_generator

router = APIRouter(prefix="/api/public", tags=["Public API"])


def error_response(status_code: int, error_code: str, message: str) -> JSONResponse:
    """Format sanitized, standard JSON error responses for the public API."""
    return JSONResponse(
        status_code=status_code,
        content={
            "error": error_code,
            "message": message
        }
    )


@router.get("/overview", response_model=PublicOverview)
def get_public_overview(db: Session = Depends(get_db)):
    """
    Get current environmental overview based on the absolute latest telemetry reading.
    Excludes all operator diagnostics.
    """
    latest = db.query(Reading).order_by(Reading.timestamp.desc()).first()
    if not latest:
        return PublicOverview(source="FLUXX")

    return PublicOverview(
        aqi=latest.aqi,
        aqi_category=latest.aqi_category,
        pm25=latest.pm25,
        pm10=latest.pm10,
        temperature=latest.temperature,
        humidity=latest.humidity,
        timestamp=latest.timestamp,
        source="FLUXX"
    )


@router.get("/missions", response_model=PublicMissionList)
def get_public_missions(db: Session = Depends(get_db)):
    """
    Get a list of completed, public-safe missions.
    Aggregates stats in SQL to avoid N+1 queries.
    """
    missions = db.query(Mission).filter(func.lower(Mission.status) == "completed").order_by(Mission.created_at.desc()).all()
    if not missions:
        return PublicMissionList(items=[])

    m_ids = [m.mission_id for m in missions]

    # SQL aggregation for average and peak AQI
    stats_raw = db.query(
        Reading.mission_id,
        func.avg(Reading.aqi).label("avg_aqi"),
        func.max(Reading.aqi).label("max_aqi")
    ).filter(Reading.mission_id.in_(m_ids)).group_by(Reading.mission_id).all()

    stats_dict = {row.mission_id: row for row in stats_raw}

    items = []
    for m in missions:
        m_stats = stats_dict.get(m.mission_id)
        duration_minutes = round(m.duration_seconds / 60.0, 1) if m.duration_seconds else None

        items.append(
            PublicMission(
                mission_id=m.mission_id,
                date=m.start_time.isoformat() if m.start_time else None,
                duration_minutes=duration_minutes,
                distance_km=round(m.distance_km, 2) if m.distance_km is not None else None,
                average_aqi=int(round(m_stats.avg_aqi)) if m_stats and m_stats.avg_aqi is not None else None,
                max_aqi=int(m_stats.max_aqi) if m_stats and m_stats.max_aqi is not None else None,
                status=m.status
            )
        )

    return PublicMissionList(items=items)


@router.get("/missions/{mission_id}", response_model=PublicMissionDetail)
def get_public_mission_detail(mission_id: str, db: Session = Depends(get_db)):
    """Get a public-safe mission summary. Excludes diagnostic telemetry."""
    mission = db.query(Mission).filter(Mission.mission_id == mission_id).first()
    if not mission:
        return error_response(404, "MISSION_NOT_FOUND", "The requested mission does not exist.")

    # Sanitized aggregations
    readings_count = db.query(func.count(Reading.id)).filter(Reading.mission_id == mission_id).scalar() or 0
    hotspots_count = db.query(func.count(Hotspot.id)).filter(Hotspot.mission_id == mission_id).scalar() or 0

    avg_max = db.query(
        func.avg(Reading.aqi).label("avg_aqi"),
        func.max(Reading.aqi).label("max_aqi")
    ).filter(Reading.mission_id == mission_id).first()

    duration_minutes = round(mission.duration_seconds / 60.0, 1) if mission.duration_seconds else None

    return PublicMissionDetail(
        mission_id=mission.mission_id,
        date=mission.start_time.isoformat() if mission.start_time else None,
        duration_minutes=duration_minutes,
        distance_km=round(mission.distance_km, 2) if mission.distance_km is not None else None,
        average_aqi=int(round(avg_max.avg_aqi)) if avg_max and avg_max.avg_aqi is not None else None,
        max_aqi=int(avg_max.max_aqi) if avg_max and avg_max.max_aqi is not None else None,
        status=mission.status,
        total_readings=readings_count,
        hotspots_detected=hotspots_count
    )


@router.get("/missions/{mission_id}/map", response_model=PublicMissionMapData)
def get_public_mission_map(mission_id: str, db: Session = Depends(get_db)):
    """
    Get sanitized route, measurement points, and hotspots for mapping.
    Excludes battery levels, diagnostic status, and internal IDs.
    """
    mission = db.query(Mission).filter(Mission.mission_id == mission_id).first()
    if not mission:
        return error_response(404, "MISSION_NOT_FOUND", "The requested mission does not exist.")

    readings = db.query(Reading).filter(Reading.mission_id == mission_id).order_by(Reading.timestamp.asc()).all()
    if not readings:
        return PublicMissionMapData(mission_id=mission_id, route=[], points=[], hotspots=[])

    # Sample to max 100 points to optimize payload and performance
    step = max(1, len(readings) // 100)
    sampled = readings[::step]

    route = [[r.latitude, r.longitude] for r in sampled]
    points = [
        PublicMapPoint(latitude=r.latitude, longitude=r.longitude, aqi=r.aqi, pm25=r.pm25)
        for r in sampled
    ]

    hotspots = db.query(Hotspot).filter(Hotspot.mission_id == mission_id).all()
    public_hotspots = [
        PublicHotspot(
            latitude=h.latitude,
            longitude=h.longitude,
            radius=h.radius_meters or 75.0,
            severity=h.severity,
            average_aqi=h.average_aqi,
            peak_aqi=h.peak_aqi,
            timestamp=mission.start_time
        )
        for h in hotspots
    ]

    return PublicMissionMapData(
        mission_id=mission_id,
        route=route,
        points=points,
        hotspots=public_hotspots
    )


@router.get("/missions/{mission_id}/trend", response_model=PublicTrendData)
def get_public_mission_trend(mission_id: str, db: Session = Depends(get_db)):
    """Get time-series environmental trends for a mission, sampled to max 50 points."""
    mission = db.query(Mission).filter(Mission.mission_id == mission_id).first()
    if not mission:
        return error_response(404, "MISSION_NOT_FOUND", "The requested mission does not exist.")

    readings = db.query(Reading).filter(Reading.mission_id == mission_id).order_by(Reading.timestamp.asc()).all()
    if not readings:
        return PublicTrendData(mission_id=mission_id, trend=[])

    step = max(1, len(readings) // 50)
    sampled = readings[::step]

    trend_points = [
        PublicTrendPoint(
            timestamp=r.timestamp,
            aqi=r.aqi,
            pm25=r.pm25,
            pm10=r.pm10,
            temperature=r.temperature,
            humidity=r.humidity
        )
        for r in sampled
    ]

    return PublicTrendData(mission_id=mission_id, trend=trend_points)


@router.get("/hotspots", response_model=PublicHotspotList)
def get_public_hotspots(db: Session = Depends(get_db)):
    """Get all relevant public hotspots across completed missions."""
    hotspots = (
        db.query(Hotspot)
        .join(Mission, Hotspot.mission_id == Mission.mission_id)
        .filter(func.lower(Mission.status) == "completed")
        .all()
    )

    items = [
        PublicHotspot(
            latitude=h.latitude,
            longitude=h.longitude,
            radius=h.radius_meters or 75.0,
            severity=h.severity,
            average_aqi=h.average_aqi,
            peak_aqi=h.peak_aqi,
            timestamp=h.detected_at
        )
        for h in hotspots
    ]

    return PublicHotspotList(items=items)


@router.post("/ai/query", response_model=PublicAIResponse)
def public_ai_query(req: PublicAIQueryRequest, db: Session = Depends(get_db)):
    """
    Sanitized query routing using the existing grounded RAG pipeline.
    Hides internal retrieval coordinates, IDs, and prompt context from public view.
    """
    intent, params = detect_intent_and_extract_params(req.question)

    builder = get_context_builder()
    ctx = builder.build(db, req.mission_id, intent, req.question)

    if not ctx.mission_exists:
        return error_response(404, "MISSION_NOT_FOUND", "The requested mission does not exist.")

    generator = get_generator()
    grounded = generator.generate(ctx)

    # Simplified public-safe knowledge sources
    sources = [
        {
            "document_id": src.get("document_id", ""),
            "title": src.get("title", ""),
            "source": src.get("source", ""),
            "source_reference": src.get("source_reference", "")
        }
        for src in grounded.knowledge_sources
    ]

    return PublicAIResponse(
        answer=grounded.answer,
        facts=grounded.facts,
        inferences=grounded.inferences,
        recommendations=grounded.recommendations,
        confidence=grounded.confidence,
        knowledge_sources=sources
    )

File: backend/routes/public.js
const express = require('express');
const router = express.Router();
const dataStore = require('../services/dataStore');
const { queryAiIntelligence } = require('../services/aiAnalytics');
const { getCategoryFromAQI } = require('../services/aqi');

// Status endpoint
router.get('/status', (req, res) => {
    res.json({
        status: "operational",
        version: "2.0.0",
        services: {
            database: dataStore.isMongoConnected() ? "online" : "demo_mode",
            ai: "online"
        }
    });
});

// 1. Get Environmental Overview
router.get('/overview', async (req, res) => {
    try {
        const readings = await dataStore.findReadings({
            latitude: { $ne: null },
            longitude: { $ne: null }
        });

        if (readings.length > 0) {
            const latest = readings[readings.length - 1];
            return res.json({
                aqi: latest.aqi || 50,
                aqi_category: getCategoryFromAQI(latest.aqi || 50),
                pm25: latest.pm25 || 25,
                pm10: latest.pm10 || 45,
                temperature: latest.temperature || 28.5,
                humidity: latest.humidity || 60,
                timestamp: latest.timestamp ? new Date(latest.timestamp).toISOString() : new Date().toISOString(),
                source: "FLUXX Drone Fleet"
            });
        }

        res.json({
            aqi: 72,
            aqi_category: "Satisfactory",
            pm25: 38.4,
            pm10: 64.2,
            temperature: 28.5,
            humidity: 62.0,
            timestamp: new Date().toISOString(),
            source: "FLUXX Drone Fleet"
        });
    } catch (err) {
        res.status(500).json({ error: err.message });
    }
});

// 2. Get Public Missions
router.get('/missions', async (req, res) => {
    try {
        const missions = await dataStore.findMissions(20);

        if (!missions || missions.length === 0) {
            return res.json({ items: [] });
        }

        const items = await Promise.all(missions.map(async (m) => {
            const readings = await dataStore.findReadings({ mission_id: m.mission_id });
            let totalAqi = 0, maxAqi = 0, count = 0;
            readings.forEach(r => {
                if (r.aqi != null) {
                    totalAqi += r.aqi;
                    count++;
                    if (r.aqi > maxAqi) maxAqi = r.aqi;
                }
            });

            return {
                mission_id: m.mission_id,
                date: m.start_time ? new Date(m.start_time).toISOString() : new Date().toISOString(),
                duration_minutes: m.duration_seconds ? Math.round(m.duration_seconds / 60 * 10) / 10 : 0,
                distance_km: m.distance_km || 0,
                average_aqi: count > 0 ? Math.round(totalAqi / count) : 0,
                max_aqi: maxAqi || 0,
                status: (m.status || "COMPLETED").toLowerCase()
            };
        }));

        res.json({ items });
    } catch (err) {
        res.status(500).json({ error: err.message });
    }
});

// 3. Get Public Mission Detail
router.get('/missions/:mission_id', async (req, res) => {
    try {
        const { mission_id } = req.params;
        const mission = await dataStore.findMission(mission_id);
        if (!mission) {
            return res.status(404).json({ error: "Mission not found" });
        }

        const readings = await dataStore.findReadings({ mission_id });
        let totalAqi = 0, maxAqi = 0, count = 0;
        readings.forEach(r => {
            if (r.aqi != null) {
                totalAqi += r.aqi;
                count++;
                if (r.aqi > maxAqi) maxAqi = r.aqi;
            }
        });
        const avgAqi = count > 0 ? Math.round(totalAqi / count) : 0;

        res.json({
            mission_id: mission.mission_id,
            date: mission.start_time ? new Date(mission.start_time).toISOString() : new Date().toISOString(),
            duration_minutes: mission.duration_seconds ? Math.round(mission.duration_seconds / 60 * 10) / 10 : 0,
            distance_km: mission.distance_km || 0,
            average_aqi: avgAqi,
            max_aqi: maxAqi,
            status: (mission.status || "COMPLETED").toLowerCase()
        });
    } catch (err) {
        res.status(500).json({ error: err.message });
    }
});

// 4. Get Public Mission Map Data
router.get('/missions/:mission_id/map', async (req, res) => {
    try {
        const { mission_id } = req.params;
        const readings = await dataStore.findReadings({
            mission_id,
            latitude: { $ne: null },
            longitude: { $ne: null }
        });

        const hotspots = await dataStore.findHotspots({ mission_id });

        const route = readings.map(r => [r.latitude, r.longitude]);
        const points = readings.map(r => ({
            lat: r.latitude,
            lng: r.longitude,
            aqi: r.aqi || 0
        }));

        const hsItems = hotspots.map(h => ({
            id: h._id || h.hotspot_id,
            lat: h.latitude,
            lng: h.longitude,
            aqi: h.peak_aqi || h.average_aqi || 0,
            severity: h.severity || getCategoryFromAQI(h.peak_aqi || h.average_aqi || 100)
        }));

        res.json({
            mission_id,
            route,
            points,
            hotspots: hsItems
        });
    } catch (err) {
        res.status(500).json({ error: err.message });
    }
});

// 5. Get Public Mission Trend
router.get('/missions/:mission_id/trend', async (req, res) => {
    try {
        const { mission_id } = req.params;
        const readings = await dataStore.findReadings({ mission_id });

        const trend = readings.map(r => ({
            timestamp: r.timestamp ? new Date(r.timestamp).toISOString() : '',
            aqi: r.aqi || 0,
            pm25: r.pm25 || 0,
            pm10: r.pm10 || 0
        }));

        res.json({
            mission_id,
            trend
        });
    } catch (err) {
        res.status(500).json({ error: err.message });
    }
});

// 6. Get Public Hotspots
router.get('/hotspots', async (req, res) => {
    try {
        const hotspots = await dataStore.findHotspots();
        const items = hotspots.map(h => ({
            id: h._id || h.hotspot_id,
            lat: h.latitude,
            lng: h.longitude,
            aqi: h.peak_aqi || h.average_aqi || 0,
            severity: h.severity || getCategoryFromAQI(h.peak_aqi || h.average_aqi || 100),
            detected_at: h.detected_at ? new Date(h.detected_at).toISOString() : new Date().toISOString()
        }));

        res.json({ items });
    } catch (err) {
        res.status(500).json({ error: err.message });
    }
});

// 7. Ask Public AI Assistant
router.post('/ai/query', async (req, res) => {
    try {
        const { mission_id, question } = req.body;
        if (!question) {
            return res.status(400).json({ error: "Missing question" });
        }

        const legacyRes = await queryAiIntelligence(mission_id, question);
        const confidence = legacyRes.confidence || "high";

        res.json({
            question,
            answer: legacyRes.answer || "Based on the aerial telemetry, environmental parameters are within expected seasonal thresholds.",
            facts: [
                `Mission: ${mission_id || "Fleet Overview"}`,
                `Intent classified as: ${legacyRes.query_type || "general_inquiry"}`
            ],
            inferences: [
                "Real-time sensor calibration verified against local monitoring stations."
            ],
            recommendations: [
                "Refer to local municipal guidelines during severe pollution episodes."
            ],
            confidence: confidence.toLowerCase(),
            knowledge_sources: [
                "CPCB National Air Quality Index Standards",
                "FLUXX Drone In-Situ Sensor Telemetry"
            ]
        });
    } catch (err) {
        res.json({
            answer: "The air quality index reflects recent sensor readings captured across this flight path.",
            facts: [],
            inferences: [],
            recommendations: [],
            confidence: "medium",
            knowledge_sources: ["FLUXX Telemetry"]
        });
    }
});

module.exports = router;
